### PR TITLE
improve performance of unrefing ancient pubkeys

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
When we combine slots into ancient append vecs, we must make sure to unref any accounts that exist in both the source slot AND the ancient append vec. To do this, we need to find the intersection of 2 sets of pubkeys. Since the same ancient appendvec is appended to by many slots, it makes sense to build a hashset of its pubkeys once and then add to it and re-use it for subsequent combine operations.
The current implementation is very slow since it iterates the ancient append vec once per slot that is combined.

#### Summary of Changes

Keep a local hashset of the pubkeys already in the current ancient appendvec.

Fixes #

### CODE REVIEW

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -3724,23 +3724,30 @@ impl AccountsDb {
         })
     }
 
+    /// 'accounts' are about to be appended to an ancient append vec. That ancient append vec may already have some accounts.
     /// unref each account in 'accounts' that already exists in 'ancient_store'
```
Comment (apfitzge) [inline runtime/src/accounts_db.rs:3728]: `ancient_store` is no longer an argument here, can you update this comment?

File: runtime/src/accounts_db.rs [hunk 2]
```diff
@@ -3750,14 +3757,9 @@ impl AccountsDb {
         ancient_store: &Arc<AccountStorageEntry>,
         accounts: &AccountsToStore,
         storage_selector: StorageSelector,
-        unref_if_already_exists: bool,
     ) -> StoreAccountsTiming {
         let (accounts, hashes) = accounts.get(storage_selector);
 
-        if unref_if_already_exists {
-            self.unref_accounts_already_in_storage(accounts, ancient_store);
-        }
```
[the remark this answers is not in the document]
Comment (jeffwashington) [reply runtime/src/accounts_db.rs:3759]: Can you elaborate, please? I do not understand.

File: runtime/src/accounts_db.rs [hunk 3]
```diff
@@ -3934,7 +3957,6 @@ impl AccountsDb {
                 &to_store,
                 StorageSelector::Primary,
                 // we are adding accounts to an existing append vec from a different slot. We need to unref each account that exists already in 'ancient_store'.
```
[the remark this answers is not in the document]
Comment (jeffwashington) [reply runtime/src/accounts_db.rs:3959]: queued up for follow-on. The current keeping track of duplicates is very expensive.

**Review by apfitzge — APPROVED** [review]
lgtm, only a nitpick comment

File: runtime/src/accounts_db.rs (hunk 1 above)
Comment (jeffwashington) [reply runtime/src/accounts_db.rs:3728]: added in follow on pr

File: runtime/src/accounts_db.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeffwashington) [reply runtime/src/accounts_db.rs:3759]: Before this change the unref was happening, just more slowly.
Before that the unref wasn't happening at all, which was a bug ;-)
There is no test for this yet. I'm getting there. Once there is a reasonable expectation that this is all correct after days of runtime on test validators and the performance is reasonable, I can lock this down with tests. In the meantime, this is only executed behind a cli arg. I think it is stabilizing.

**Pull request merged**